### PR TITLE
Add Scoring Pallet

### PR DESCRIPTION
**Pull request by roynrishingha**

### CODE REVIEW

**Comment by amarts** [comment]
```
278 |             let identifier = Ss58Identifier::to_scoring_id(&(&journal.digest).encode()[..])
    |                                                             ^^^^^^^^^^^^^^^^^ help: change this to: `journal.digest`
```

**Comment by roynrishingha** [comment]
> ```
> 278 |             let identifier = Ss58Identifier::to_scoring_id(&(&journal.digest).encode()[..])
>     |                                                             ^^^^^^^^^^^^^^^^^ help: change this to: `journal.digest`
> ```

Ya. I saw. Clippy fails. Will fix it today.

File: pallets/scoring/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,82 @@
+[package]
+name = "pallet-scoring"
```
Comment (amarts) [inline pallets/scoring/Cargo.toml:2]: I have a fundamental question. Should we name it with a noun? (ie, **pallet-score** or action driven name is fine? ie, what is present here **pallet-scoring**.

File: pallets/scoring/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+description = "Manage network scores"
+version = "0.8.1"
```
Comment (amarts) [inline pallets/scoring/Cargo.toml]: Version PR is now merged, and hence '0.9.0'?

File: pallets/scoring/Readme.md [hunk 3]
```diff
@@ -0,0 +1 @@
+#test
```
Comment (amarts) [inline pallets/scoring/Readme.md:1]: recommended file name for Readme is `README.md` (good to stick to that as a practise).

**Review by amarts — COMMENTED** [review]
minor comments

File: pallets/scoring/Cargo.toml (hunk 1 above)
Comment (rohibhan) [reply pallets/scoring/Cargo.toml:2]: Scoring is good i think

File: pallets/scoring/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+			journal: RatingInputOf<T>,
+			authorization: AuthorizationIdOf,
+		) -> DispatchResult {
+			let author = <T as Config>::EnsureOrigin::ensure_origin(origin)?.subject();
+
+			let registry_id =
+				pallet_registry::Pallet::<T>::is_a_delegate(&authorization, author.clone(), None)
+					.map_err(<pallet_registry::Error<T>>::from)?;
+
+			ensure!(
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs:248]: We can remove the use constant here. All we need is a value above 0.

File: pallets/scoring/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+				!<JournalHashes<T>>::contains_key(journal.digest),
+				Error::<T>::DigestAlreadyAnchored
+			);
+
+			// let id_digest = <T as frame_system::Config>::Hashing::hash(
+			// 	&[&(journal.digest).encode()[..]].concat()[..],
+			// );
+
+			let identifier = Ss58Identifier::to_scoring_id(&journal.digest.encode()[..])
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs]: Identifier creation logic needs to be fixed to maintain the uniqueness.

File: pallets/scoring/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+		RatingCannotBeZero,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		///
+		/// Create a new rating identifier and associates it with its
+		/// controller. The controller (issuer) is the owner of the identifier.
+		///
+		/// * origin: the identity of the Transaction Author. Transaction author
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs]: remove the pay transaction fees

File: pallets/scoring/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+		///   pays the transaction fees
+		/// * tx_journal: the incoming rating entry.
+		/// * `authorization`: The authorization ID of the delegate who is
+		///   allowed to perform this action.
+		#[pallet::call_index(0)]
+		#[pallet::weight({0})]
+		pub fn entries(
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs]: entry / add_rating?

File: pallets/scoring/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn journal_hashes)]
+	pub type JournalHashes<T> =
+		StorageMap<_, Blake2_128Concat, RatingEntryHashOf<T>, (), ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn tid_entries)]
+	pub type TidEntries<T> = StorageDoubleMap<
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs]: change this to TransactionIdentifiers

File: pallets/scoring/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+			);
+
+			Self::deposit_event(Event::JournalEntry {
+				identifier,
+				entity: journal.entry.entity.clone(),
+				author,
+			});
+
+			Ok(())
+		}
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs:292]: Need to add a couple of more pallet calls with a CouncilOrigin. 
- Remove an entity and its aggregate score from storage
- Remove the rating entries once an entity aggregate is removed. 
This can be a separate PR after the merge. @NiranjanAP fyi

File: pallets/scoring/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+	pub use cord_primitives::{curi::Ss58Identifier, CountOf, RatingOf};
+	use cord_utilities::traits::CallSources;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_std::{prelude::Clone, str};
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	/// Registry Identifier
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs:65]: Remove unused ones, if any

File: pallets/scoring/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+			entity: EntityIdentifierOf<T>,
+			author: RatingCreatorIdOf<T>,
+		},
+		/// Aggregate scores has been updated.
+		/// \[entity identifier\]
+		AggregateUpdated { entity: EntityIdentifierOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs:200]: Remove unused ones, if any

File: pallets/scoring/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+				RatingEntryOf::<T> {
+					entry: journal.entry.clone(),
+					digest: journal.digest,
+					created_at: block_number,
+					registry: registry_id,
+					creator: author.clone(),
+				},
+			);
+			<JournalHashes<T>>::insert(journal.digest, ());
+			<TidEntries<T>>::insert(
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs]: Storage name change

File: pallets/scoring/src/mock.rs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+	pub enum Test where
+		Block = frame_system::mocking::MockBlock<Test>,
+		NodeBlock = frame_system::mocking::MockBlock<Test>,
+		UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Test>,
+	{
+		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
+		Schema:pallet_schema::{Pallet, Call, Storage, Event<T>},
+		Registry: pallet_registry::{Pallet, Storage, Call,Event<T>},
+		Scoring: pallet_scoring::{Pallet, Call, Storage, Event<T>},
+		Balances: pallet_balances::{Pallet, Call, Storage, Event<T>},
```
Comment (smohan-dw) [inline pallets/scoring/src/mock.rs]: remove balances

File: pallets/scoring/src/mock.rs [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+	type AccountData = pallet_balances::AccountData<u128>;
+	type OnNewAccount = ();
+	type OnKilledAccount = ();
+	type SystemWeightInfo = ();
+	type SS58Prefix = SS58Prefix;
+	type OnSetCode = ();
+	type MaxConsumers = ConstU32<2>;
+}
+
+impl pallet_balances::Config for Test {
```
Comment (smohan-dw) [inline pallets/scoring/src/mock.rs]: remove

File: pallets/scoring/src/types.rs [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+	///entity Identifier
+	pub entity: EntityIdentifierOf,
+	/// transaction identifier for which the score is requested
+	pub tid: TransactionIdentfierOf,
+	///score collector identifier
+	pub collector: CollectorIdentifierOf,
+	/// score type
+	pub rating_type: RatingTypeOf, //overall, delivery
+	///entity rating
+	pub rating: RatingOf, //u32 15,16
```
Comment (smohan-dw) [inline pallets/scoring/src/types.rs]: // u32 15,16 ?

File: pallets/scoring/src/types.rs [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+	RatingEntryDetails,
+	RatingEntryHashOf,
+	BlockNumber,
+	RegistryIdOf,
+	RatingCreatorIdOf,
+> {
+	///journal entry
+	pub entry: RatingEntryDetails,
+	/// tx digest
+	pub digest: RatingEntryHashOf,
```
Comment (smohan-dw) [inline pallets/scoring/src/types.rs:82]: Double storage of digest. DIgest is also present inside the entry.

File: pallets/scoring/src/types.rs [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+	/// The block number in which journal entry is included
+	pub created_at: BlockNumber,
+	/// Registry Identifier
+	pub registry: RegistryIdOf,
+	/// Rating creator.
+	pub creator: RatingCreatorIdOf,
```
Comment (smohan-dw) [inline pallets/scoring/src/types.rs:88]: same with the creator also. It's present inside entry

File: runtime/src/lib.rs [hunk 18]
```diff
@@ -772,6 +772,20 @@ parameter_types! {
...
 }
 
+impl pallet_scoring::Config for Runtime {
+	type RatingCreatorIdOf = DidIdentifier;
+	type EnsureOrigin = pallet_did::EnsureDidOrigin<DidIdentifier, AccountId>;
+	type OriginSuccess = pallet_did::DidRawOrigin<AccountId, DidIdentifier>;
+	type RuntimeEvent = RuntimeEvent;
+	type MinScoreValue = MinScoreValue;
+	type WeightInfo = weights::pallet_scoring::WeightInfo<Runtime>;
+	type ValueLimit = ConstU32<72>;
```
Comment (smohan-dw) [inline runtime/src/lib.rs:781]: where is this used?

File: runtime/src/lib.rs [hunk 19]
```diff
@@ -772,6 +772,20 @@ parameter_types! {
 	pub const MaxEncodedLength: u32 = 15_360;
 }
 
+impl pallet_scoring::Config for Runtime {
+	type RatingCreatorIdOf = DidIdentifier;
+	type EnsureOrigin = pallet_did::EnsureDidOrigin<DidIdentifier, AccountId>;
+	type OriginSuccess = pallet_did::DidRawOrigin<AccountId, DidIdentifier>;
+	type RuntimeEvent = RuntimeEvent;
+	type MinScoreValue = MinScoreValue;
```
Comment (smohan-dw) [inline runtime/src/lib.rs]: this can be removed

File: runtime/src/lib.rs [hunk 20]
```diff
@@ -955,6 +972,7 @@ mod benches {
 		[pallet_did_names, DidNames]
 		[pallet_network_membership, NetworkMembership]
 		[pallet_sudo, Sudo]
+		[pallet_scoring, Scoring]
```
Comment (smohan-dw) [inline runtime/src/lib.rs:969]: put it above sudo. Keep sudo as the last pallet

File: scripts/run_benches_for_pallets.sh [hunk 21]
```diff
@@ -115,6 +116,9 @@ for PALLET in "${PALLETS[@]}"; do
   pallet_unique)
     FOLDER="unique"
     ;;
+  pallet_scoring)
+    FOLDER="scoring"
+    ;;
 
   *)
     # Exit early.
```
Comment (smohan-dw) [inline scripts/run_benches_for_pallets.sh:124]: Missing scoring benchmark entry on run_benches_for_runtime.sh

File: pallets/scoring/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,347 @@
...
+		EntityIdentifierOf<T>,
+		Blake2_128Concat,
+		RatingTypeOf,
+		ScoreEntryOf,
+		OptionQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn journal_hashes)]
+	pub type JournalHashes<T> =
```
Comment (smohan-dw) [inline pallets/scoring/src/lib.rs]: Remove this storage

File: pallets/scoring/src/lib.rs (hunk 9 above)
Comment (rohibhan) [reply pallets/scoring/src/lib.rs:292]: Noted @smohan-dw. As we talked about, we will take this one in the Next PR @NiranjanAP.

File: runtime/src/lib.rs (hunk 18 above)
Comment (rohibhan) [reply runtime/src/lib.rs:781]: pub type TransactionIdentifierOf<T> = BoundedVec<u8, <T as Config>::ValueLimit>;

**Review by smohan-dw — DISMISSED** [review]
LGTM

**Review by smohan-dw — APPROVED** [review]
LGTM

**Review by amarts — APPROVED** [review]
